src/extension_commands.ml

```diff
@@ -950,6 +950,185 @@ module Search_by_type = struct
   ;;
 end
 
+module Navigate_holes = struct
+  let extension_name = "Navigate between Typed Holes"
+
+  let ocaml_lsp_doesnt_support_typed_holes ocaml_lsp =
+    not (Ocaml_lsp.can_handle_typed_holes ocaml_lsp)
+  ;;
+
+  let is_valid_text_doc textdoc =
+    match TextDocument.languageId textdoc with
+    | "ocaml" | "ocaml.interface" | "reason" -> true
+    | _ -> false
+  ;;
+
+  let send_request_to_lsp client doc =
+    let uri = TextDocument.uri doc in
+    Custom_requests.send_request client Custom_requests.typedHoles uri
+  ;;
+
+  let show_selection selection text_editor =
+    TextEditor.set_selection text_editor selection;
+    TextEditor.revealRange
+      text_editor
+      ~range:(Selection.to_range selection)
+      ~revealType:TextEditorRevealType.InCenterIfOutsideViewport
+      ()
+  ;;
+
+  let jump_to_range range text_editor =
+    let open Promise.Syntax in
+    let+ _ =
+      Window.showTextDocument
+        ~document:(TextEditor.document text_editor)
+        ~preserveFocus:true
+        ()
+    in
+    let selection =
+      let anchor = Range.start range in
+      let active = Range.end_ range in
+      Selection.makePositions ~anchor ~active
+    in
+    show_selection selection text_editor
+  ;;
+
+  module QuickPickItemWithRange = struct
+    type t =
+      { item : QuickPickItem.t
+      ; range : Range.t
+      }
+
+    let t_of_js js =
+      let range = Ojs.get_prop_ascii js "pl_range" |> Range.t_of_js in
+      let item = QuickPickItem.t_of_js js in
+      { item; range }
+    ;;
+
+    let t_to_js t =
+      let item = QuickPickItem.t_to_js t.item in
+      Ojs.set_prop_ascii item "pl_range" (Range.t_to_js t.range);
+      item
+    ;;
+  end
+
+  module QuickPick = Vscode.QuickPick.Make (QuickPickItemWithRange)
+
+  let display_results (results : Range.t list) text_editor client instance =
+    let open Promise.Syntax in
+    let selected_item = ref false in
+    let text_document = TextEditor.document text_editor in
+    let quickPickItems =
+      List.map results ~f:(fun range ->
+        let line = Position.line @@ Range.end_ range in
+        let item =
+          QuickPickItem.create
+            ~label:(Printf.sprintf "Line %d" line)
+            ~description:(TextLine.text @@ TextDocument.lineAt ~line text_document)
+            ()
+        in
+        { QuickPickItemWithRange.item; range })
+    in
+    let quickPick =
+      QuickPick.set
+        (Window.createQuickPick (module QuickPickItemWithRange) ())
+        ~title:"Typed Holes"
+        ~activeItems:[]
+        ~busy:false
+        ~enabled:true
+        ~placeholder:"Use arrow keys to preview / Select to jump to it"
+        ~selectedItems:[]
+        ~ignoreFocusOut:true
```

Comment:
```suggestion
        ~ignoreFocusOut:false
```

This option is very disruptive for the user's flow, we should only enable it when there is a real need for it. We discussed it at some point, but I think it's not needed in the end isn't it ?